.evergreen/socks5srv.py

```diff
@@ -0,0 +1,237 @@
+
+    For example, if there is a mapping (localhost, 1000) to (localhost, 2000),
+    then this also adds (127.0.0.1, 1000) to (localhost, 2000)."""
+
+    for src, dst in self.mappings:
+      host, port = src
+      try:
+        addrs = socket.getaddrinfo(host, port, socket.AF_UNSPEC, socket.SOCK_STREAM)
+      except socket.gaierror:
+        continue
+
+      existing_src_entries = [src for src, dst in self.mappings]
+      for af, socktype, proto, canonname, sa in addrs:
+        if af == socket.AF_INET and sa not in existing_src_entries:
+          self.mappings.append((sa, dst))
+        elif af == socket.AF_INET6 and sa[:2] not in existing_src_entries:
+          self.mappings.append((sa[:2], dst))
+
+  def remap(self, hostport):
+    """Re-map a (host, port) tuple to a new (host, port) tuple if that was requested"""
+
+    for src, dst in self.mappings:
+      if hostport == src:
+        return dst
+    return hostport
+
+class Socks5Server(socketserver.ThreadingTCPServer):
+  """A simple Socks5 proxy server"""
+
+  def __init__(self, server_address, RequestHandlerClass, args):
+    socketserver.ThreadingTCPServer.__init__(self,
+                                             server_address,
+                                             RequestHandlerClass)
+    self.args = args
+    self.address_remapper = AddressRemapper(args.map)
+
+class Socks5Handler(socketserver.BaseRequestHandler):
+  """Request handler for Socks5 connections"""
+
+  def finish(self):
+    """Called after handle(), always just closes the connection"""
+
+    self.request.close()
+
+  def read_exact(self, n):
+    """Read n bytes from a socket
+
+    In Socks5, strings are prefixed with a single byte containing
+    their length. This method reads a bytes string containing n bytes
+    (where n can be a number or a bytes object containing that
+    single byte).
+
+    If reading from the client ends prematurely, this returns None.
+    """
+
+    if type(n) is bytes:
+      if len(n) == 0:
+        return None
+      assert len(n) == 1
+      n = n[0]
+    result = b''
+    while len(result) < n:
+      buf = self.request.recv(n - len(result))
+      if buf == b'':
+        return None
+      result += buf
+    return result
+
+  def create_outgoing_tcp_connection(self, dst, port):
+    """Create an outgoing TCP connection to dst:port"""
+
+    outgoing = None
+    for res in socket.getaddrinfo(dst, port, socket.AF_UNSPEC, socket.SOCK_STREAM):
+      af, socktype, proto, canonname, sa = res
+      try:
+        outgoing = socket.socket(af, socktype, proto)
+      except OSError as msg:
+        continue
+      try:
+        outgoing.connect(sa)
+      except OSError as msg:
+        outgoing.close()
+        continue
+      break
+    return outgoing
+
+  def handle(self):
+    """Handle the Socks5 communication with a freshly connected client"""
+
+    # Client greeting
+    if self.request.recv(1) != b'\x05': # Socks5 only
```

Comment:
Any request.recv/sendall call can raise an OSError exception, do you want to handle those explicitly or just let the thread fail with an exception? I think the latter is fine.

Reply:
Yeah, I’d also say so – I don’t really see a better alternative and I think this behavior is fine for our usage.